File: partial-refund-tax-miscalculated/node/flag-creditmemo-tax-mismatch.js
/**
 * Flag Magento 2 or Adobe Commerce credit memos whose tax was computed from
 * the full order instead of the refunded items.
 *
 * Magento's credit memo tax totals collector is supposed to prorate tax per
 * line item using qty being refunded versus qty invoiced. Long standing bugs,
 * seen across magento2 GitHub issues 8797, 9929, 10982, 14713, 23938, 32222,
 * and 34586, instead cause it to copy the order's full tax_amount and
 * base_tax_amount onto the credit memo, notably when the credit memo is
 * created from the admin order view, when multiple partial credit memos are
 * issued against the same invoice, or when the display currency differs from
 * the base currency. CreditmemoItemInterface.tax_amount is a snapshot stored
 * at creation time, never re-derived later, so a wrong number stays wrong
 * forever.
 *
 * This script never edits an existing credit memo, since Magento has no
 * supported REST write for that. It recomputes the expected proportional tax
 * from the order's own item data, compares it to each credit memo's reported
 * base_tax_amount, and writes a reconciliation report. Only under an
 * explicit DRY_RUN=false does it optionally POST a new corrective refund
 * call carrying an adjustment_positive or adjustment_negative argument. In
 * dry run it only prints the proposed payload. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/magento/partial-refund-tax-miscalculated/
 */
import { pathToFileURL } from "node:url";

const MAGENTO_URL = (process.env.MAGENTO_URL || "https://example.test").replace(/\/$/, "");
const ADMIN_USERNAME = process.env.MAGENTO_ADMIN_USERNAME || "admin";
const ADMIN_PASSWORD = process.env.MAGENTO_ADMIN_PASSWORD || "change-me";
const ADMIN_TOKEN = process.env.MAGENTO_ADMIN_TOKEN || "";
const ORDER_IDS = (process.env.ORDER_IDS || "").split(",").map((o) => o.trim()).filter(Boolean);
const TAX_EPSILON = Number(process.env.TAX_EPSILON || 0.01);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * Pure decision function. Takes only primitive numeric inputs already
 * extracted from the order/creditmemo JSON and returns a plain object for
 * the caller to log or act on. No I/O.
 */
export function isCreditMemoTaxMismatched(orderItemTaxAmount, orderItemQtyOrdered, creditMemoItemQty, creditMemoBaseTaxAmount, epsilon = TAX_EPSILON) {
  const expectedTax = orderItemQtyOrdered
    ? orderItemTaxAmount * (creditMemoItemQty / orderItemQtyOrdered)
    : 0;
  const delta = creditMemoBaseTaxAmount - expectedTax;
  return { expectedTax, delta, mismatched: Math.abs(delta) > epsilon };
}

/**
 * Sum the pure per-line expected tax across every refunded line on a credit
 * memo, using each order item's own tax_amount and qty_ordered.
 */
export function expectedTaxForCreditMemo(orderItemsById, creditMemo) {
  let expectedTotal = 0;
  for (const cmItem of creditMemo.items || []) {
    const orderItem = orderItemsById[cmItem.order_item_id];
    if (!orderItem) continue;
    expectedTotal += isCreditMemoTaxMismatched(
      orderItem.tax_amount || 0,
      orderItem.qty_ordered || 0,
      cmItem.qty || 0,
      0,
    ).expectedTax;
  }
  return expectedTotal;
}

export function buildAdjustmentPayload(delta) {
  if (delta > 0) return { arguments: { adjustment_negative: Math.round(delta * 100) / 100 } };
  return { arguments: { adjustment_positive: Math.round(Math.abs(delta) * 100) / 100 } };
}

async function getToken() {
  if (ADMIN_TOKEN) return ADMIN_TOKEN;
  const res = await fetch(`${MAGENTO_URL}/rest/V1/integration/admin/token`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ username: ADMIN_USERNAME, password: ADMIN_PASSWORD }),
  });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

async function getOrder(token, orderId) {
  const res = await fetch(`${MAGENTO_URL}/rest/V1/orders/${orderId}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

async function getCreditmemosForOrder(token, orderId) {
  const params = new URLSearchParams({
    "searchCriteria[filterGroups][0][filters][0][field]": "order_id",
    "searchCriteria[filterGroups][0][filters][0][value]": orderId,
    "searchCriteria[filterGroups][0][filters][0][conditionType]": "eq",
    "searchCriteria[pageSize]": "100",
  });
  const res = await fetch(`${MAGENTO_URL}/rest/V1/creditmemo?${params}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  const body = await res.json();
  return body.items || [];
}

async function applyAdjustment(token, orderId, delta) {
  const payload = buildAdjustmentPayload(delta);
  const res = await fetch(`${MAGENTO_URL}/rest/V1/order/${orderId}/refund`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throw new Error(`Magento ${res.status}`);
  return res.json();
}

export async function run() {
  const token = await getToken();
  const flagged = [];

  for (const orderId of ORDER_IDS) {
    const order = await getOrder(token, orderId);
    const orderItemsById = {};
    for (const item of order.items || []) orderItemsById[item.item_id] = item;
    const creditmemos = await getCreditmemosForOrder(token, orderId);

    for (const cm of creditmemos) {
      const expectedTax = expectedTaxForCreditMemo(orderItemsById, cm);
      const actualTax = cm.base_tax_amount || 0;
      const delta = actualTax - expectedTax;
      const mismatched = Math.abs(delta) > TAX_EPSILON;
      if (!mismatched) continue;

      const row = {
        order_increment_id: order.increment_id,
        creditmemo_increment_id: cm.increment_id,
        expected_tax: Math.round(expectedTax * 10000) / 10000,
        actual_tax: Math.round(actualTax * 10000) / 10000,
        delta: Math.round(delta * 10000) / 10000,
      };
      flagged.push(row);
      console.warn(`Order ${row.order_increment_id} creditmemo ${row.creditmemo_increment_id}: expected_tax=${row.expected_tax} actual_tax=${row.actual_tax} delta=${row.delta}`);

      const payload = buildAdjustmentPayload(delta);
      console.log("Proposed adjustment payload:", JSON.stringify(payload));
      if (!DRY_RUN) await applyAdjustment(token, orderId, delta);
    }
  }

  console.log(`Done. ${flagged.length} creditmemo(s) flagged, ${DRY_RUN ? "dry run, nothing written" : "corrective refund attempted where flagged"}.`);
  return flagged;
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: partial-refund-tax-miscalculated/python/flag_creditmemo_tax_mismatch.py
"""Flag Magento 2 or Adobe Commerce credit memos whose tax was computed from the
full order instead of the refunded items.

Magento's credit memo tax totals collector is supposed to prorate tax per line
item using qty being refunded versus qty invoiced. Long standing bugs, seen
across magento2 GitHub issues 8797, 9929, 10982, 14713, 23938, 32222, and
34586, instead cause it to copy the order's full tax_amount and
base_tax_amount onto the credit memo, notably when the credit memo is created
from the admin order view, when multiple partial credit memos are issued
against the same invoice, or when the display currency differs from the base
currency. CreditmemoItemInterface.tax_amount is a snapshot stored at creation
time, never re-derived later, so a wrong number stays wrong forever.

This script never edits an existing credit memo, since Magento has no
supported REST write for that. It recomputes the expected proportional tax
from the order's own item data, compares it to each credit memo's reported
base_tax_amount, and writes a reconciliation report. Only under an explicit
DRY_RUN=false does it optionally POST a new corrective refund call carrying
an adjustment_positive or adjustment_negative argument. In dry run it only
prints the proposed payload. Safe to run again and again.

Guide: https://www.allanninal.dev/magento/partial-refund-tax-miscalculated/
"""
import os
import csv
import logging
import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("flag_creditmemo_tax_mismatch")

MAGENTO_URL = os.environ.get("MAGENTO_URL", "https://example.test").rstrip("/")
ADMIN_USERNAME = os.environ.get("MAGENTO_ADMIN_USERNAME", "admin")
ADMIN_PASSWORD = os.environ.get("MAGENTO_ADMIN_PASSWORD", "change-me")
ADMIN_TOKEN = os.environ.get("MAGENTO_ADMIN_TOKEN")
ORDER_IDS = [o.strip() for o in os.environ.get("ORDER_IDS", "").split(",") if o.strip()]
TAX_EPSILON = float(os.environ.get("TAX_EPSILON", "0.01"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
OUTPUT_CSV = os.environ.get("OUTPUT_CSV", "creditmemo_tax_mismatches.csv")


def get_token():
    if ADMIN_TOKEN:
        return ADMIN_TOKEN
    r = requests.post(
        f"{MAGENTO_URL}/rest/V1/integration/admin/token",
        json={"username": ADMIN_USERNAME, "password": ADMIN_PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def get_order(token, order_id):
    r = requests.get(
        f"{MAGENTO_URL}/rest/V1/orders/{order_id}",
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def get_creditmemos_for_order(token, order_id):
    params = {
        "searchCriteria[filterGroups][0][filters][0][field]": "order_id",
        "searchCriteria[filterGroups][0][filters][0][value]": order_id,
        "searchCriteria[filterGroups][0][filters][0][conditionType]": "eq",
        "searchCriteria[pageSize]": 100,
    }
    r = requests.get(
        f"{MAGENTO_URL}/rest/V1/creditmemo",
        params=params,
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json().get("items", [])


def is_creditmemo_tax_mismatched(order_item_tax_amount, order_item_qty_ordered,
                                  creditmemo_item_qty, creditmemo_base_tax_amount, epsilon=TAX_EPSILON):
    """Pure decision function. Takes only primitive numeric inputs already
    extracted from the order/creditmemo JSON and returns a plain dict for the
    caller to log or act on. No I/O.
    """
    if not order_item_qty_ordered:
        expected_tax = 0.0
    else:
        expected_tax = order_item_tax_amount * (creditmemo_item_qty / order_item_qty_ordered)
    delta = creditmemo_base_tax_amount - expected_tax
    return {
        "expectedTax": expected_tax,
        "delta": delta,
        "mismatched": abs(delta) > epsilon,
    }


def expected_tax_for_creditmemo(order_items_by_id, creditmemo):
    """Sum the pure per-line expected tax across every refunded line on a
    credit memo, using each order item's own tax_amount and qty_ordered.
    """
    expected_total = 0.0
    for cm_item in creditmemo.get("items", []):
        order_item = order_items_by_id.get(cm_item.get("order_item_id"))
        if not order_item:
            continue
        expected_total += is_creditmemo_tax_mismatched(
            order_item.get("tax_amount", 0) or 0,
            order_item.get("qty_ordered", 0) or 0,
            cm_item.get("qty", 0) or 0,
            0,  # only expectedTax is used here, actual comparison happens at the creditmemo level
        )["expectedTax"]
    return expected_total


def build_adjustment_payload(delta):
    if delta > 0:
        return {"arguments": {"adjustment_negative": round(delta, 2)}}
    return {"arguments": {"adjustment_positive": round(abs(delta), 2)}}


def apply_adjustment(token, order_id, delta):
    payload = build_adjustment_payload(delta)
    r = requests.post(
        f"{MAGENTO_URL}/rest/V1/order/{order_id}/refund",
        json=payload,
        headers={"Authorization": f"Bearer {token}"},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()


def run():
    token = get_token()
    flagged = []
    for order_id in ORDER_IDS:
        order = get_order(token, order_id)
        order_items_by_id = {item.get("item_id"): item for item in order.get("items", [])}
        creditmemos = get_creditmemos_for_order(token, order_id)

        for cm in creditmemos:
            expected_tax = expected_tax_for_creditmemo(order_items_by_id, cm)
            actual_tax = cm.get("base_tax_amount", 0) or 0
            delta = actual_tax - expected_tax
            mismatched = abs(delta) > TAX_EPSILON
            if not mismatched:
                continue

            row = {
                "order_increment_id": order.get("increment_id"),
                "creditmemo_increment_id": cm.get("increment_id"),
                "expected_tax": round(expected_tax, 4),
                "actual_tax": round(actual_tax, 4),
                "delta": round(delta, 4),
            }
            flagged.append(row)
            log.warning(
                "Order %s creditmemo %s: expected_tax=%s actual_tax=%s delta=%s",
                row["order_increment_id"], row["creditmemo_increment_id"],
                row["expected_tax"], row["actual_tax"], row["delta"],
            )

            payload = build_adjustment_payload(delta)
            log.info("Proposed adjustment payload: %s", payload)
            if not DRY_RUN:
                apply_adjustment(token, order_id, delta)

    if flagged:
        with open(OUTPUT_CSV, "w", newline="") as fh:
            writer = csv.DictWriter(fh, fieldnames=[
                "order_increment_id", "creditmemo_increment_id",
                "expected_tax", "actual_tax", "delta",
            ])
            writer.writeheader()
            writer.writerows(flagged)

    log.info("Done. %d creditmemo(s) flagged, %s.", len(flagged),
              "dry run, nothing written" if DRY_RUN else "corrective refund attempted where flagged")
    return flagged


if __name__ == "__main__":
    run()
